### PR TITLE
feature: Add mysql failed query logging + fixed queries that break ONLY_FULL_GROUP_BY

### PR DESCRIPTION
**Pull request by laf**
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you signed the [Contributors agreement](http://docs.librenms.org/General/Contributing/)
- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

Will inline comment on these. Genuinely shouldn't break anything but fixes a lot of broken queries or ones that don't work with ONLY_FULL_GROUP_BY enabled which is default.

### CODE REVIEW

File: build.sql [hunk 1]
```diff
@@ -58,7 +58,6 @@ CREATE TABLE IF NOT EXISTS `ports_vlans` ( `port_vlan_id` int(11) NOT NULL AUTO_
...
 CREATE TABLE IF NOT EXISTS `cmpMemPool` (  `cmp_id` int(11) NOT NULL auto_increment,  `Index` varchar(8) NOT NULL,  `cmpName` varchar(32) NOT NULL,  `cmpValid` varchar(8) NOT NULL,  `device_id` int(11) NOT NULL,  `cmpUsed` int(11) NOT NULL,  `cmpFree` int(11) NOT NULL,  `cmpLargestFree` int(11) NOT NULL,  `cmpAlternate` tinyint(4) default NULL,  PRIMARY KEY  (`cmp_id`),  KEY `device_id` (`device_id`)) ENGINE=MyISAM  DEFAULT CHARSET=latin1;
 CREATE TABLE IF NOT EXISTS `hrDevice` ( `hrDevice_id` int(11) NOT NULL auto_increment,  `device_id` int(11) NOT NULL,  `hrDeviceIndex` int(11) NOT NULL,  `hrDeviceDescr` text NOT NULL,  `hrDeviceType` text NOT NULL,  `hrDeviceErrors` int(11) NOT NULL,  `hrDeviceStatus` text NOT NULL,  `hrProcessorLoad` tinyint(4) default NULL,  PRIMARY KEY  (`hrDevice_id`),  KEY `device_id` (`device_id`)) ENGINE=MyISAM  DEFAULT CHARSET=latin1;
-CREATE TABLE IF NOT EXISTS `dbSchema` ( `revision` int(11) NOT NULL default '0', PRIMARY KEY (`revision`)) ENGINE=MyISAM DEFAULT CHARSET=latin1;
```
Comment (laf) [inline build.sql]: We have already created the table so no point trying again.

File: html/includes/functions.inc.php [hunk 2]
```diff
@@ -788,9 +788,9 @@ function getlocations()
 
     // Fetch regular locations
     if ($_SESSION['userlevel'] >= '5') {
-        $rows = dbFetchRows('SELECT D.device_id,location FROM devices AS D GROUP BY location ORDER BY location');
```
Comment (laf) [inline html/includes/functions.inc.php]: device_id is never used so get rid to fix query

File: html/includes/functions.inc.php [hunk 3]
```diff
@@ -788,9 +788,9 @@ function getlocations()
...
+        $rows = dbFetchRows('SELECT location FROM devices AS D GROUP BY location ORDER BY location');
     } else {
-        $rows = dbFetchRows('SELECT D.device_id,location FROM devices AS D, devices_perms AS P WHERE D.device_id = P.device_id AND P.user_id = ? GROUP BY location ORDER BY location', array($_SESSION['user_id']));
```
Comment (laf) [inline html/includes/functions.inc.php]: Same here

File: html/includes/print-alert-rules.php [hunk 4]
```diff
@@ -143,7 +143,7 @@
 
 echo '</select></td>';
 
-$count_query = 'SELECT COUNT(id)';
```
Comment (laf) [inline html/includes/print-alert-rules.php]: Can't use COUNT(id) without using GROUP BY which we don't need.

File: html/includes/print-alert-rules.php [hunk 5]
```diff
@@ -152,7 +152,7 @@
     $param = array($device['device_id']);
 }
 
-$query       = " FROM alert_rules $sql ORDER BY id ASC";
```
Comment (laf) [inline html/includes/print-alert-rules.php]: don't need ORDER BY for count query so moved to full query.

File: html/includes/print-map.inc.php [hunk 6]
```diff
@@ -84,7 +84,7 @@
                              `D1`.`device_id` != `D2`.`device_id`
                              $where
                              $sql
-                      GROUP BY `P1`.`port_id`,`P2`.`port_id`
```
Comment (laf) [inline html/includes/print-map.inc.php]: GROUP BY must contain all the fields in the SELECT. Tested and # results are the same.

File: html/includes/print-map.inc.php [hunk 7]
```diff
@@ -125,7 +125,7 @@
                              `remote_device_id` != 0
                              $where
                              $sql
-                      GROUP BY `P1`.`port_id`,`P2`.`port_id`
```
Comment (laf) [inline html/includes/print-map.inc.php]: As above.

File: html/includes/table/alert-schedule.inc.php [hunk 8]
```diff
@@ -25,7 +25,7 @@
     $sql .= " AND (`S`.`title` LIKE '%$searchPhrase%' OR `S`.`start` LIKE '%$searchPhrase%' OR `S`.`end` LIKE '%$searchPhrase%')";
 }
 
-$count_sql = "SELECT COUNT(`id`) $sql";
```
Comment (laf) [inline html/includes/table/alert-schedule.inc.php]: id doesn't exist, it's schedule_id

File: html/includes/table/customers.inc.php [hunk 9]
```diff
@@ -38,7 +38,7 @@
     $sql .= " LIMIT $limit_low,$limit_high";
 }
 
-$sql = "SELECT * $sql";
```
Comment (laf) [inline html/includes/table/customers.inc.php]: We only use port_descr_descr so change SELECT to only that.

File: html/includes/table/poll-log.inc.php [hunk 10]
```diff
@@ -26,9 +26,12 @@
     $sort = 'last_polled_timetaken DESC';
 }
 
-$sql .= " AND D.status ='1' AND D.ignore='0' AND D.disabled='0' ORDER BY $sort";
```
Comment (laf) [inline html/includes/table/poll-log.inc.php]: Same as other comment, don't need ORDER BY for count query.

File: html/pages/ports.inc.php [hunk 11]
```diff
@@ -87,9 +87,9 @@
 <?php
 
 if ($_SESSION['userlevel'] >= 5) {
-    $results = dbFetchRows("SELECT `device_id`,`hostname` FROM `devices` GROUP BY `hostname` ORDER BY `hostname`");
```
Comment (laf) [inline html/pages/ports.inc.php]: GROUP BY hostname is irrelevant here. hostnames are unique anyway. Same for other queries in this file.

File: html/pages/routing/bgp.inc.php [hunk 12]
```diff
@@ -240,14 +240,6 @@
             }
         }
 
-        $peerhost = dbFetchRow('SELECT * FROM ipaddr AS A, ports AS I, devices AS D WHERE A.addr = ? AND I.port_id = A.port_id AND D.device_id = I.device_id', array($peer['bgpPeerIdentifier']));
```
Comment (laf) [inline html/pages/routing/bgp.inc.php]: `$peername` set and never used so removed as query for `$peerhost` is totally wrong (non-existent table).

File: html/pages/search/arp.inc.php [hunk 13]
```diff
@@ -38,7 +38,7 @@
     $param[] = $_SESSION['user_id'];
 }
 
-$sql .= " WHERE M.port_id = P.port_id AND P.device_id = D.device_id $where GROUP BY `device_id` ORDER BY `hostname`";
```
Comment (laf) [inline html/pages/search/arp.inc.php]: No need for GROUP BY device_id as they are unique already.

File: html/pages/search/ipv4.inc.php [hunk 14]
```diff
@@ -35,7 +35,7 @@
     $param[] = $_SESSION['user_id'];
 }
 
-$sql .= " $where GROUP BY `hostname` ORDER BY `hostname`";
```
Comment (laf) [inline html/pages/search/ipv4.inc.php]: Same here.

File: html/pages/search/ipv6.inc.php [hunk 15]
```diff
@@ -34,7 +34,7 @@
     $param[] = $_SESSION['user_id'];
 }
 
-$sql .= " $where GROUP BY `hostname` ORDER BY `hostname`";
```
Comment (laf) [inline html/pages/search/ipv6.inc.php]: Same here.

File: html/pages/search/mac.inc.php [hunk 16]
```diff
@@ -34,7 +34,7 @@
     $param[] = $_SESSION['user_id'];
 }
 
-$sql .= " $where GROUP BY `hostname` ORDER BY `hostname`";
```
Comment (laf) [inline html/pages/search/mac.inc.php]: Same here

File: html/pages/services.inc.php [hunk 17]
```diff
@@ -99,10 +99,10 @@
     </tr>
 <?php
 if ($_SESSION['userlevel'] >= '5') {
-    $host_sql = 'SELECT * FROM devices AS D, services AS S WHERE D.device_id = S.device_id GROUP BY D.hostname ORDER BY D.hostname';
```
Comment (laf) [inline html/pages/services.inc.php]: Same here

File: includes/dbFacile.php [hunk 18]
```diff
@@ -39,23 +39,14 @@ function dbQuery($sql, $parameters = array())
         }
     }
 
-    /*
```
Comment (laf) [inline includes/dbFacile.php]: Removed redundant code

File: includes/dbFacile.php [hunk 19]
```diff
@@ -39,23 +39,14 @@ function dbQuery($sql, $parameters = array())
...
-        $time_start = microtime(true);
-     */
-
     $result = mysqli_query($database_link, $fullSql);
-    // sets $this->result
-    /*
-        if($this->logFile) {
-        $time_end = microtime(true);
-        fwrite($this->logFile, date('Y-m-d H:i:s') . "\n" . $fullSql . "\n" . number_format($time_end - $time_start, 8) . " seconds\n\n");
+    if (!$result) {
```
Comment (laf) [inline includes/dbFacile.php]: So what we do here is check if the mysql log level is INFO or ERROR AND don't contain duplicate entry in the error (we have these and they need to be fixed but will come back to that) and print out any mysql errors OR if the log level is DEBUG then just print out all failed queries.

I've then used this to find all broken queries.

File: includes/dbFacile.php [hunk 20]
```diff
@@ -420,7 +411,11 @@ function dbMakeQuery($sql, $parameters)
     krsort($namedParams);
 
     // split on question-mark and named placeholders
-    $result = preg_split('/(\?|:[a-zA-Z0-9_-]+)/', $sql, -1, (PREG_SPLIT_NO_EMPTY | PREG_SPLIT_DELIM_CAPTURE));
+    if (preg_match('/(\[\[:[\w]+:\]\])/', $sql)) {
```
Comment (laf) [inline includes/dbFacile.php]: This is a 'hack' I couldn't get a preg_split regex to work across all cases so this retains existing compatibility but checks for people using mysql regex with [[:SOMETHING:]].

If someone can find a good regex to cover all basis then go for it.

It currently splits on ? and : as they are both placeholders for variable substitution and causes problems with [[:digit:]] as per issue #5118 (so fixes that)

File: includes/defaults.inc.php [hunk 21]
```diff
@@ -31,6 +31,8 @@
 
 // MySQL extension to use
 $config['db']['extension']       = 'mysqli';//mysql and mysqli available
+// MySQL Debug level
+$config['mysql_log_level']       = 'ERROR';
```
Comment (laf) [inline includes/defaults.inc.php]: Set this as the default so all installs will log errors. We could set it to NONE but imho this is now handy to find non-working things easily and won't spam the log file anyway.

Comment (laf) [inline includes/defaults.inc.php]: Well unless someone has an issue.

File: includes/discovery/cisco-pw.inc.php [hunk 22]
```diff
@@ -53,7 +53,7 @@
     // Cycle the list of pseudowires we cached earlier and make sure we saw them again.
     foreach ($device['pws_db'] as $pw_id => $pseudowire_id) {
         if (empty($device['pws'][$pw_id])) {
-            dbDelete('vlans', '`pseudowire_id` = ?', array($pseudowire_id));
```
Comment (laf) [inline includes/discovery/cisco-pw.inc.php]: Yeah, lol

File: includes/sql-schema/update.php [hunk 23]
```diff
@@ -36,34 +36,6 @@
     $insert = 1;
 }
 
-// For transition from old system
```
Comment (laf) [inline includes/sql-schema/update.php]: I literally can't find where we do anything with revision - I think it's a hang up from the fork so got rid of it.

File: sql-schema/002.sql [hunk 24]
```diff
@@ -75,7 +75,6 @@ ALTER TABLE `temperature` ADD `temp_precision` INT(11) NULL DEFAULT '1';
 UPDATE temperature SET temp_precision=10 WHERE temp_tenths=1;
 ALTER TABLE `temperature` ADD  `temp_index` INT NOT NULL AFTER  `temp_host` , ADD  `temp_mibtype` VARCHAR( 32 ) NOT NULL AFTER  `temp_index`;
 ALTER TABLE `temperature` DROP `temp_tenths`;
-CREATE TABLE IF NOT EXISTS `dbSchema` ( `revision` int(11) NOT NULL default '0', PRIMARY KEY (`revision`)) ENGINE=MyISAM DEFAULT CHARSET=latin1;
```
Comment (laf) [inline sql-schema/002.sql]: We've already created this table (then tried 3 more times!)

**Comment by laf** [comment]
Pushed further updates to the base sql build which fixes invalid queries. I have a feeling build.sql contains a lot of what was in 001-006.sql as most of the tidying up was done there.

**Comment by laf** [comment]
Exported the database using this new branch, the old branch with both strict disabled and some defaults in centos. A diff shows the difference below (comparing old with non-strict and new shows only differences are datetime)

```
[root@mysql01 ~]# diff /tmp/ltest_old_nonstrict.sql /tmp/ltest_new.sql
151,152c151,152
<   `start` timestamp NOT NULL DEFAULT '0000-00-00 00:00:00',
<   `end` timestamp NOT NULL DEFAULT '0000-00-00 00:00:00',
---
>   `start` timestamp NOT NULL DEFAULT '1970-01-01 23:00:01',
>   `end` timestamp NOT NULL DEFAULT '1970-01-01 23:00:01',
895c895,896
<   `version` int(11) NOT NULL
---
>   `version` int(11) NOT NULL DEFAULT '0',
>   PRIMARY KEY (`version`)
1275c1276
<   `datetime` datetime NOT NULL DEFAULT '0000-00-00 00:00:00',
---
>   `datetime` datetime NOT NULL DEFAULT '1000-01-01 00:00:00',
3391c3392
< -- Dump completed on 2017-01-06 15:30:25
---
> -- Dump completed on 2017-01-06 15:29:19
[root@mysql01 ~]# diff /tmp/ltest_old_strict.sql /tmp/ltest_new.sql
142a143,168
> -- Table structure for table `alert_schedule`
> --
>
> DROP TABLE IF EXISTS `alert_schedule`;
> /*!40101 SET @saved_cs_client     = @@character_set_client */;
> /*!40101 SET character_set_client = utf8 */;
> CREATE TABLE `alert_schedule` (
>   `schedule_id` int(11) NOT NULL AUTO_INCREMENT,
>   `start` timestamp NOT NULL DEFAULT '1970-01-01 23:00:01',
>   `end` timestamp NOT NULL DEFAULT '1970-01-01 23:00:01',
>   `title` varchar(255) NOT NULL,
>   `notes` text NOT NULL,
>   PRIMARY KEY (`schedule_id`)
> ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
> /*!40101 SET character_set_client = @saved_cs_client */;
>
> --
> -- Dumping data for table `alert_schedule`
> --
>
> LOCK TABLES `alert_schedule` WRITE;
> /*!40000 ALTER TABLE `alert_schedule` DISABLE KEYS */;
> /*!40000 ALTER TABLE `alert_schedule` ENABLE KEYS */;
> UNLOCK TABLES;
>
> --
869c895,896
<   `version` int(11) NOT NULL
---
>   `version` int(11) NOT NULL DEFAULT '0',
>   PRIMARY KEY (`version`)
1238a1266,1296
> -- Table structure for table `eventlog`
> --
>
> DROP TABLE IF EXISTS `eventlog`;
> /*!40101 SET @saved_cs_client     = @@character_set_client */;
> /*!40101 SET character_set_client = utf8 */;
> CREATE TABLE `eventlog` (
>   `event_id` int(11) NOT NULL AUTO_INCREMENT,
>   `host` int(11) NOT NULL DEFAULT '0',
>   `device_id` int(11) NOT NULL,
>   `datetime` datetime NOT NULL DEFAULT '1000-01-01 00:00:00',
>   `message` text CHARACTER SET latin1,
>   `type` varchar(64) CHARACTER SET latin1 DEFAULT NULL,
>   `reference` varchar(64) CHARACTER SET latin1 NOT NULL,
>   PRIMARY KEY (`event_id`),
>   KEY `host` (`host`),
>   KEY `datetime` (`datetime`),
>   KEY `device_id` (`device_id`)
> ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;
> /*!40101 SET character_set_client = @saved_cs_client */;
>
> --
> -- Dumping data for table `eventlog`
> --
>
> LOCK TABLES `eventlog` WRITE;
> /*!40000 ALTER TABLE `eventlog` DISABLE KEYS */;
> /*!40000 ALTER TABLE `eventlog` ENABLE KEYS */;
> UNLOCK TABLES;
>
> --
3334c3392
< -- Dump completed on 2017-01-06 15:29:54
---
> -- Dump completed on 2017-01-06 15:29:19

```

**Comment by Rosiak** [comment]
Doesn't seem to break my test install, haven't reviewed the actual code yet.

**Comment by murrant** [comment]
Looks good, I didn't test and compare a fresh db run, been running it for a day without issue.   I have a few errors logged, but I'll fix those in a different PR.

**Comment by laf** [comment]
We ok for a merge then?

**Pull request merged**